### ISSUE TITLE
MacOS Build successful - what next?

### ISSUE THREAD

**Issue by estefan3112** [mmatyas/pegasus-frontend#96]
Hi, I followed your instructions and succeeded in MacOS to build the executable into /opt/pegasus-frontend -> there I have the following directory structure
- assets
- doc
- etc
Makefile
- src (including the executable pegasus-fe)
- tests
I can start pegasus-fe, but it is empty since obviously all other files that the frontend probably expects from the raspberry pi implementation are missing. What would be the next steps from there? I have a current retropie environment running on my rpi3. 
Many thanks.

**Comment by mmatyas** [comment]
Hi, at the moment Pegasus requires the configuration files of EmulationStation, so if it can't find them, or can't load them, you'll see no games. (Error messages are still on my todo list, sorry!)

First, Pegasus tries to find `es_systems.cfg`, which should be at `$HOME/.emulationstation/es_systems.cfg` or at `/etc/emulationstation/es_systems.cfg`. Then it tries to load the `gamelist.xml` of each system, which should be under the platform's `<path>` directory, or under `$HOME/.emulationstation/gamelists/<name>/` or `/etc/emulationstation/gamelists/<name>/`.

If the files have been found or there've been any errors, you'll see that on the console output or in the log file `lastrun.log`, which I think on Mac should be either at `$HOME/.config/pegasus-frontend/` or `"~/Library/Preferences/pegasus-frontend`.

**Comment by estefan3112** [comment]
Success! And looks good already on my Mac. What I did so far:
1. copied entire content of $HOME/.emulationstation/ from my pi to the Mac
2. copied entire content of /etc/emulationstation/ from my pi to the Mac (I also used the /etc folder and allowed copy with admin password)
3. changed the N64 command in es_systems.cfg to mupen command on the Mac
4. adjusted some paths in the N64 gamelixt.xml to the Mac environment just to see whether it works

And it works like a charm. Mupen64 starts smoothly.

![pegasus-mac](https://user-images.githubusercontent.com/24915559/30631183-aa27a546-9de3-11e7-8d76-8d2ae92a3519.png)

Pegasus log all fine, just one error:
qrc:/themes/pegasus-grid/GameGrid.qml:28: TypeError: Cannot read property 'currentGameIndex' of null
Need to look into that separately.

Only Joypad seems a mess, as always in the Mac environment...

**Comment by mmatyas** [comment]
Great! Good to hear it works. The error is a harmless bug that occurs when you quit Pegasus; I've fixed that yesterday, it just haven't made into the daily release yet.

I'm not familiar with the gamepad handling on Mac, but if it's a Pegasus-related bug (eg. you can't reconfigure some buttons), feel free to open an issue for it.

**Comment by estefan3112** [comment]
In the meantime I also got one Retroarch core running, nice. Looks really promising.

Regarding joypads I have this problem with all of my sticks: Example PS4 via USB. A Mac program called Joystick Show (http://chibatacreations.com/app/joystickshow/) tells me that the joystick is properly recognized as 'Joystick #0 : Wireless Controller' with all the buttons. Pegasus-FE sees it as Gamepad #1, however, only the A,B,X,Y buttons and shoulder left, shoulder right buttons flash green when pressed. Everything else does not seem to work. What to do next? Thanks!

**Comment by mmatyas** [comment]
Do the start/pause/center buttons work? It sounds like the gamepad axes aren't detected properly; do they work in Joystick Show or, for example, [here](http://html5gamepad.com/)? Do you have any other devices to try?

**Comment by estefan3112** [comment]
So I did further tests with both my PS4 Controller (USB) and my Logitech F310 (X-Box Mode with X-Box MacOS software drivers): On this website, both Joypads are fully recognized (Google Chrome, not Safari) with all buttons working.
Pegasus-fe: PS4 shows only the four buttons and the two shoulder buttons, nothing else. Logitech F310 is not recognized at all, i.e. no Controller seen. 
I am currently waiting for a 8bitDO NES30 Arcade Stick. 
All in all it seems to me that pegasus-fe has a problem in my environment to properly find joypads.
Cheers!

**Comment by mmatyas** [comment]
It sounds like a bug in one of the Qt modules, the library Pegasus uses for cross-platform input handling. Do you have Qt Creator installed? There's an example application there (you can build and run it similarly to Pegasus), if the controllers don't work in it either, we might want to report it to the Qt devs. Here's where you can find it in Qt Creator:

![kepernyokep_2017-09-23_16-18-13_2](https://user-images.githubusercontent.com/4354863/30773983-e73d2ab0-a07b-11e7-9770-0e6575b3838c.png)

**Comment by estefan3112** [comment]
I think I have a problem with qt: I have Qt 5.9.1 installed via Homebrew and qt Creator via Homebrew-Cask. Although Qt Creator sees Qt 5.9.1, there are no Examples available because there is no Qt visible. I double checked the Preferences and there is nothing else I can do about it. This is a pity
![qt creator](https://user-images.githubusercontent.com/24915559/31045025-220c04a4-a5db-11e7-960a-e835695e7568.png)
 because qt via Homebrew is so easy to manage.

**Comment by estefan3112** [comment]
So just to let you know that I installed QT5.9.1 (official offline installer) and ran the QTGamepad Example. Same result: Only the six buttons described above work in my PS4 Gamepad (via USB). I will give the RumblePad a try and update this post then.
![qtgamepad](https://user-images.githubusercontent.com/24915559/31049872-755a3060-a63d-11e7-81c5-d0fb6d966072.png)

**Comment by mmatyas** [comment]
Thanks for the testing, it does seem like a bug in Qt on Mac. Looking around their issue tracker, I've actually found reports of gamepad issues on Mac that might be relevant:

- https://bugreports.qt.io/browse/QTBUG-55788
- https://bugreports.qt.io/browse/QTBUG-62128

In the first one Qt seems to be unable to use gamepads at all, while in the second it just doesn't see the gamepads. To see additional debug information like in the reports, you can define the environmental variable `QT_LOGGING_RULES="qt.gamepad=true"` before starting Pegasus. You should see `qt.gamepad: Available backends: ("darwin")` on the terminal, and *maybe* some additional, useful message.

Unfortunately there doesn't seem to be much activity on the reports, so until then we're also blocked :/ If you want to create a new issue by the way, [here's a guide](https://wiki.qt.io/Reporting_Bugs) for it (the Project is "Qt", the Affected version is "5.9.0" and the Component is "Gamepad").

**Comment by estefan3112** [comment]
Hi - I tried my best with the environment variable, but failed (as a non-coder).
So I added a note under QTBUG-62128. Now it is probably time to wait...

**Comment by mmatyas** [comment]
The gamepad handling has been updated recently, which should fix the issues on Mac too. If you have some time, could you check if it works correctly for you too?

**Comment by estefan3112** [comment]
Yes, the controller works now! I think this can be closed now. Thank you. 
PS: Now struggling to launch any games...but I need to dig deeper into it...
